### PR TITLE
[WIP] [FR] add Sequential.append, .extend , .insert & .pop

### PR DESCRIPTION
**Pull request by codexetreme**
adding features as required from : #20117 
cc: @SsnL @apaszke

### CODE REVIEW

**Comment by codexetreme** [comment]
Why does it show 220 changed files, I have changed only 1 in reality

**Comment by codexetreme** [comment]
Here, how will the functions be implemented? In Python only or in c++, such what when, say `append()` is called, it has some c++ bound function that is called

File: torch/nn/modules/container.py [hunk 1]
```diff
@@ -87,6 +87,19 @@ def __dir__(self):
         keys = [key for key in keys if not key.isdigit()]
         return keys
 
+    # need to add code, just a starter
+    def append(self):
```
Comment (SsnL) [inline torch/nn/modules/container.py]: all these need to be on sequential class.

**Review by ssnl — COMMENTED** [review]
python sequential and cpp sequential are unrelated. the cpp one seems to have a different API, so I think you should focus on the python one for now. 

cpp: https://github.com/pytorch/pytorch/blob/7bc8562a9a43af7ae779fdef22326efe8644ae92/torch/csrc/api/include/torch/nn/modules/sequential.h

**Comment by codexetreme** [comment]
Alright, I will work on it accordingly

File: torch/nn/modules/container.py [hunk 2]
```diff
@@ -87,6 +87,32 @@ def __dir__(self):
...
+    def extend(self,sequential):
+        prev_len = self.__len__()
+        if isinstance(sequential,Sequential):
+            for idx,layer in enumerate(sequential):
+                self.add_module(str(idx + prev_len),layer)
+        else:
+            raise ValueError('The given parameter {}, is of type {},\n expected it to be of type torch.nn.Sequential'.format(sequential,torch.typename(sequential)))
+
+    def insert(self,i,layer):
+        pass
```
Comment (codexetreme) [inline torch/nn/modules/container.py]: For insert, since it is a ordered dict, there are some values that will be in strings that are non numeric, so how exactly can the additional layers be added? eg: say the sequential container has 2 layers `test` and `test2` and the input of `i` is 1, then where exactly  do we add it?

**Comment by codexetreme** [comment]
Here, where to add the tests?

File: torch/nn/modules/container.py (hunk 2 above)
Comment (crcrpar) [reply torch/nn/modules/container.py]: I'm not sure this is OK for core developers, but the below might insert a layer at index `i`.
Note that I didn't test this snippet.

```python
    def insert(self, i, new_layer):
        new_modules = OrderedDict()
        for j, (key, layer) in enumerate(self._modules.items()):
            if i == j:
                new_modules[str(i)] = new_layer
            new_modules[key] = layer
        del self._modules
        self._modules = new_modules
```

File: torch/nn/modules/container.py [hunk 3]
```diff
@@ -87,6 +87,33 @@ def __dir__(self):
         keys = [key for key in keys if not key.isdigit()]
         return keys
 
+    # need to add code, just a starter
```
Comment (crcrpar) [inline torch/nn/modules/container.py]: Could you remove this line, please?

**Comment by crcrpar** [comment]
> Here, where to add the tests?

At least, add tests to `test_nn.py` and my suggestion is around https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L1040.

File: torch/nn/modules/container.py (hunk 2 above)
Comment (codexetreme) [reply torch/nn/modules/container.py]: Ok, I will implement similarly, if the Core Devs aren't okay, I can change it :)

**Comment by codexetreme** [comment]
@crcrpar Can I just make a new test method? Since this new method will test for all possible cases like popping on an empty sequential, and so on.. (the method you linked `test_named_modules` seems to be for names modules specifically, whereas these methods are an extension) and I feel should be tested in a method of their own to check many possible cases.

**Comment by crcrpar** [comment]
Yes, we should add new tests because methods you are implementing are new.
I wanted to suggest the location of the new tests by the link because test
files are huge.

**Comment by codexetreme** [comment]
@crcrpar I have added the tests in a separate function as requested. Also, I added the insert functionality, but the code seems to be very hacky (since there is a need to create a new OrderedDict plus I change keys if they clash when inserting) One possibility is to maybe not have insert functionality? because from how I normally use pytorch, I would just create a new sequential and make it "insert" like so


```
def test():

    l1 = nn.Linear(2,3)
    l2 = nn.Linear(3,3)
    l3 = nn.Linear(3,3)
    l4 = nn.Linear(3,1)
    k = nn.Sequential(l1,l2,l4)
    mid_layer = nn.Sequential(l3)

    p = k[0:1](torch.tensor([2,3],dtype=torch.float))
    
    # basically I have "inserted" my layer here 
    p = mid_layer(p)
    
    p = k[-1](p)
    print (p)
```

**Comment by crcrpar** [comment]
@SsnL @apaszke Can I get your opinion about `insert` method?

IMO, one difference between PyTorch's `Sequential` and Chainer's is that the former inherits `Module` while the latter inherits `ChainList` which maintains children layers using a Python list.

**Comment by codexetreme** [comment]
any updates on this?

**Comment by codexetreme** [comment]
Can anyone please give an update this? This seems to be an easy issue to fix. It'll reduce the open PR count too :)

**Comment by jamaliki** [comment]
Hi, is there something I can contribute to on this? I was thinking of making a PR for these features but then I discovered this open one. If there is anything to add, please let me know.
